price-feeder/oracle/provider/websocket_controller.go

```diff
@@ -0,0 +1,189 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+)
+
+type (
+	MessageHandler func(int, []byte)
+
+	// WebsocketController defines a provider agnostic websocket handler
+	// that manages reconnecting, subscribing, and receiving messages
+	WebsocketController struct {
+		ctx              context.Context
+		providerName     Name
+		websocketURL     url.URL
+		subscriptionMsgs []interface{}
+		messageHandler   MessageHandler
+		logger           zerolog.Logger
+		client           *websocket.Conn
+		mu               sync.Mutex
```

Comment:
Just a suggestion to indicate what the mtx is protecting

```suggestion
		
		mtx               sync.Mutex
		client           *websocket.Conn
```